redfish-core/lib/assembly.hpp

```diff
@@ -18,6 +21,94 @@
 namespace redfish
 {
 
+constexpr std::array<std::string_view, 9> chassisAssemblyInterfaces = {
+    "xyz.openbmc_project.Inventory.Item.Vrm",
+    "xyz.openbmc_project.Inventory.Item.Tpm",
+    "xyz.openbmc_project.Inventory.Item.Panel",
+    "xyz.openbmc_project.Inventory.Item.Battery",
+    "xyz.openbmc_project.Inventory.Item.DiskBackplane",
+    "xyz.openbmc_project.Inventory.Item.Board",
+    "xyz.openbmc_project.Inventory.Item.Connector",
+    "xyz.openbmc_project.Inventory.Item.Drive",
+    "xyz.openbmc_project.Inventory.Item.Board.Motherboard"};
+
+inline void doGetAssociatedChassisAssembly(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& chassisPath,
+    std::function<void(const std::vector<std::string>& assemblyList)>&&
+        callback)
+{
+    BMCWEB_LOG_DEBUG("Get associated chassis assembly");
+
+    sdbusplus::message::object_path endpointPath{chassisPath};
+    endpointPath /= "assembly";
+
+    dbus::utility::getAssociatedSubTreePaths(
+        endpointPath,
+        sdbusplus::message::object_path("/xyz/openbmc_project/inventory"), 0,
+        chassisAssemblyInterfaces,
+        [asyncResp, chassisPath, callback{callback}](
+            const boost::system::error_code& ec,
+            const dbus::utility::MapperGetSubTreePathsResponse& subtreePaths) {
+        if (ec)
+        {
+            if (ec.value() != EBADR)
+            {
+                BMCWEB_LOG_ERROR(
+                    "DBUS response error for getAssociatedSubTreePaths {}",
+                    ec.value());
+                messages::internalError(asyncResp->res);
+                return;
+            }
+            callback(std::vector<std::string>());
+            return;
+        }
+
+        std::vector<std::string> sortedAssemblyList = subtreePaths;
+        std::ranges::sort(sortedAssemblyList);
+
+        callback(sortedAssemblyList);
+    });
+}
+
+/**
+ * @brief Get chassis path with given chassis ID
+ * @param[in] asyncResp - Shared pointer for asynchronous calls.
+ * @param[in] chassisID - Chassis to which the assemblies are
+ * associated.
+ * @param[in] callback
+ *
+ * @return None.
+ */
+inline void getChassisAssembly(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& chassisID,
+    std::function<void(const std::optional<std::string>& validChassisPath,
+                       const std::vector<std::string>& sortedAssemblyList)>&&
+        callback)
+{
+    BMCWEB_LOG_DEBUG("Get ChassisAssembly");
+
+    // get the chassis path
+    redfish::chassis_utils::getValidChassisPath(
+        asyncResp, chassisID,
+        [asyncResp, callback{callback}](
+            const std::optional<std::string>& validChassisPath) {
+        if (!validChassisPath)
```

Comment:
Maybe a comment here would be helpful.

Seems like we are doing the callback whether we had a validChassisPath or not. Is the goal for the callback to handle the error in a more meaningful way? The one caller added isn't differentiating between the cases.

Or, would it be more clear to handle the !validChassisPath here as an error and not call the callback at all?

Reply:
I'm making a function `getChassisAssembly()` which will be used in several places (currently in assembly, but it will be used other places), and I'd like to make it so that the semantical error handling (not dbus-error) is done at the caller.

The function `redfish::chassis_utils::getValidChassisPath()` is designed to return `validChassisPath` (which may also represent non-existence)  which is also considered as semantical error and thus it will be passed to the caller.

The caller will process those conditions as errors, not by `getChassisAssembly()`. So it is passing this to the caller.